R/do_azureml_parallel.R

```diff
@@ -0,0 +1,252 @@
+
+  node_count <- 1L
+  process_count_per_node <- 1L
+  r_env <- NULL
+  max_run_duration_seconds <- NULL
+
+  if (!is.null(obj$args$job_timeout)) {
+    max_run_duration_seconds <- obj$args$job_timeout
+    obj$args$job_timeout <- NULL
+  }
+
+  if (!is.null(obj$args$node_count)) {
+    node_count <- obj$args$node_count
+    obj$args$node_count <- NULL
+  }
+
+  if (!is.null(obj$args$process_count_per_node)) {
+    process_count_per_node <- obj$args$process_count_per_node
+    obj$args$process_count_per_node <- NULL
+  }
+
+  if (!is.null(obj$args$r_env)) {
+    r_env <- obj$args$r_env
+    obj$args$r_env <- NULL
+  }
+
+  it <- iterators::iter(obj)
+  args_list <- as.list(it)
+
+  exportenv <- tryCatch({
+    qargs <- quote(list(...))
+    args <- eval(qargs, envir)
+    environment(do.call(.make_dots_env, args))
+  },
+  error = function(e) {
+    new.env(parent = emptyenv())
+  })
+  noexport <- union(obj$noexport, obj$argnames)
+  packages <- foreach::getexports(expr, exportenv, envir, bad = noexport)
+  packages <- c(packages, obj$packages)
+  vars <- ls(exportenv)
+
+  export <- unique(obj$export)
+  ignore <- intersect(export, vars)
+  if (length(ignore) > 0) {
+    export <- setdiff(export, ignore)
+  }
+
+  # add explicitly exported variables to exportenv
+  if (length(export) > 0) {
+    if (obj$verbose)
+      cat(sprintf("explicitly exporting variables(s): %s\n",
+                  paste(export, collapse = ", ")))
+
+    for (sym in export) {
+      if (!exists(sym, envir, inherits = TRUE))
+        stop(sprintf('unable to find variable "%s"', sym))
+
+      val <- get(sym, envir, inherits = TRUE)
+      assign(sym, val, pos = exportenv, inherits = FALSE)
+    }
+  }
+
+  assign("packages", obj$packages, global_env)
+
+  expr <- compiler::compile(expr, env = envir,
+                            options = list(suppressUndefined = TRUE))
+
+  assign("expr", expr, global_env)
+  assign("exportenv", exportenv, global_env)
+
+  # divide args into MPI tasks
+  task_args <- split_tasks(args_list, node_count, process_count_per_node)
+  assign("task_args", task_args, global_env)
+
+  source_dir <- paste0("foreach_run_", as.integer(Sys.time()))
+  dir.create(source_dir)
+
+  saveRDS(global_env, file = file.path(source_dir, "env.rds"))
+
+  # submit estimator job to the cluster
+  generate_entry_script(source_directory = source_dir)
+  est <- estimator(source_directory = source_dir,
+                   compute_target = data$cls,
+                   entry_script = "entry_script.py",
+                   environment = r_env,
+                   max_run_duration_seconds = max_run_duration_seconds)
+
+  dist_backend <- azureml$core$runconfig$MpiConfiguration()
+  dist_backend$process_count_per_node <- process_count_per_node
+
+  run_config <- est$run_config
+  run_config$framework <- "python"
+  run_config$communicator <- "IntelMpi"
+  run_config$mpi <- dist_backend
+  run_config$node_count <- node_count
+
+  experiment_name <- "r-foreach"
```

Comment:
@heemanshusuri since this will show in in run history UI, should we give users the ability to optionally specify the experiment name?

Reply:
added.